worlds/cv64/entrances.py

```diff
@@ -0,0 +1,151 @@
+    ename.tosci_end: {"connection": ("next", rname.tower_of_science)},
+    # Tower of Sorcery
+    ename.tosor_start: {"connection": ("prev", rname.tower_of_sorcery)},
+    ename.tosor_end: {"connection": ("next", rname.tower_of_sorcery)},
+    # Room of Clocks
+    ename.roc_gate: {"connection": ("next", rname.room_of_clocks)},
+    # Clock Tower
+    ename.ct_to_door1: {"connection": rname.ct_middle, "rule": iname.clocktower_key1},
+    ename.ct_from_door1: {"connection": rname.ct_start, "rule": iname.clocktower_key1},
+    ename.ct_to_door2: {"connection": rname.ct_end, "rule": iname.clocktower_key2},
+    ename.ct_from_door2: {"connection": rname.ct_middle, "rule": iname.clocktower_key2},
+    ename.ct_renon: {"connection": rname.renon, "add conds": ["shopsanity"]},
+    ename.ct_door_3: {"connection": ("next", rname.clock_tower), "rule": iname.clocktower_key3},
+    # Castle Keep
+    ename.ck_slope_jump: {"connection": rname.roc_main, "add conds": ["hard"]},
+    ename.ck_drac_door: {"connection": rname.ck_drac_chamber, "rule": "Dracula"}
+}
+
+add_conds = {"carrie": ("carrie_logic", True, True),
+             "hard": ("hard_logic", True, True),
+             "not hard": ("hard_logic", False, True),
+             "shopsanity": ("shopsanity", True, True)}
+
+stage_connection_types = {"prev": "end region",
+                          "next": "start region",
+                          "alt": "start region"}
+
+
+def get_entrance_info(entrance: str, info: str) -> Union[str, Tuple[str, str], List[str], None]:
+    if info in entrance_info[entrance]:
+        return entrance_info[entrance][info]
+    return None
+
+
+def get_warp_entrances(active_warp_list: List[str]) -> Dict[str, str]:
+    # Create the starting stage Entrance.
+    warp_entrances = {get_stage_info(active_warp_list[0], "start region"): "Start stage"}
+
+    # Create the warp Entrances.
+    for i in range(1, len(active_warp_list)):
+        mid_stage_region = get_stage_info(active_warp_list[i], "mid region")
+        warp_entrances.update({mid_stage_region: f"Warp {i}"})
+
+    return warp_entrances
+
+
+def verify_entrances(options: CV64Options, entrances: List[str],
+                     active_stage_exits: Dict[str, Dict[str, Union[str, int, None]]]) -> Dict[str, str]:
+    verified_entrances = {}
+
+    for ent_name in entrances:
+        ent_add_conds = get_entrance_info(ent_name, "add conds")
+
+        # Check any options that might be associated with the Entrance before adding it.
+        add_it = True
+        if ent_add_conds is not None:
+            for cond in ent_add_conds:
+                if not ((getattr(options, add_conds[cond][0]).value == add_conds[cond][1]) == add_conds[cond][2]):
+                    add_it = False
+
+        if not add_it:
+            continue
+
+        # Add the Entrance to the verified entrances if the above check passes.
+        connection = get_entrance_info(ent_name, "connection")
```

Comment:
is it fine for this to be None? it looks very problematic

Reply:
At the moment, there are no entrances in entrance_info without a "connection" key so it shouldn't ever return None in the first place. But I can add a check for it just in case.

Reply:
the important bit and why i asked is because the method itself is written (and typed) such that it can return None